src/controllers/auth.controller.js

```
@@ -0,0 +1,128 @@
+
+const EMAIL_PATTERN = /^[\w.+-]+@([\w-]+\.){1,3}[\w-]{2,}$/;
+const USERNAME_PATTERN = /^[a-zA-Z\- ]{3,20}$/;
+
+function validateUsername(value) {
+  if (!value) return 'Username is required';
+  if (!USERNAME_PATTERN.test(value)) return 'Username is not valid';
+}
+
+function validateEmail(value) {
+  if (!value) return 'Email is required';
+  if (!EMAIL_PATTERN.test(value)) return 'Email is not valid';
+}
+
+function validatePassword(value) {
+  if (!value) return 'Password is required';
+  if (value.length < 6) return 'At least 6 characters';
+}
+
+const register = async (req, res, next) => {
+  const { username, email, password } = req.body;
+
+  const errors = {
+    username: validateUsername(),
+    email: validateEmail(),
+    password: validatePassword(),
+  };
+
+  if (!errors.email || !errors.password || !errors.username) {
+    throw ApiError.badRequest('Bad request', errors);
+  }
+
+  const hashedPass = await bcrypt.hash(password, 10);
+
+  await userService.register(username, email, hashedPass);
+  res.send({ message: 'OK' });
+};
+
+const activate = async (req, res) => {
+  const { activationToken } = req.params;
+  const user = await User.findOne({ where: { activationToken } });
+
+  if (!user) {
+    res.sendStatus(404);
+    return;
+  }
+  user.activationToken = null;
+  user.save();
+
+  res.send(user);
+};
+
+async function generateTokens(res, user) {
+  const normalizedUser = userService.normalize(user);
+
+  const accessToken = jwtService.sign(normalizedUser);
+  const refreshAccessToken = jwtService.signRefresh(normalizedUser);
+
+  await tokenService.save(normalizedUser.id, refreshAccessToken);
+  res.cookie('refreshToken', refreshAccessToken, {
+    maxAge: 30 * 24 * 60 * 60 * 1000,
+    httpOnly: true,
+  });
+
+  res.send({
+    user: normalizedUser,
+    accessToken,
+  });
+}
+
+const login = async (req, res) => {
+  const { email, password } = req.body;
+
+  const user = await userService.findByEmail(email);
```

The task requires that if a user tries to log in with a non-activated account, you should inform them to activate their email. After fetching the user, you need to check if `user.activationToken` is not `null`. If it's not `null`, the user is not yet active, and you should return an error instead of proceeding to check the password.